tests/unit/src/test/scala/tests/DebugDiscoverySuite.scala

```diff
@@ -0,0 +1,241 @@
+          "run"
+        ).toJson
+      )
+      _ <- debugger.initialize
+      _ <- debugger.launch
+      _ <- debugger.configurationDone
+      _ <- debugger.shutdown
+      output <- debugger.allOutput
+    } yield assertNoDiff(output, "what about grapes?")
+  }
+
+  test("no-main") {
+    for {
+      _ <- server.initialize(
+        s"""/metals.json
+           |{
+           |  "a": {}
+           |}
+           |/a/src/main/scala/a/Main.scala
+           |package a
+           |object Stuff {
+           |  val stuff = "you guessed it, stuff"
+           |  System.exit(0)
+           |}
+           |""".stripMargin
+      )
+      _ <- server.didOpen("a/src/main/scala/a/Main.scala")
+      result <- server
+        .startDebuggingUnresolved(
+          new DebugFullyUnresolvedParams(
+            s"file://${workspace}/a/src/main/scala/a/Main.scala",
+            "run"
+          ).toJson
+        )
+        .recover { case e: BuildTargetContainsNoMainException =>
+          e
+        }
+    } yield assertNoDiff(
+      result.toString,
+      BuildTargetContainsNoMainException("a").toString()
+    )
+  }
+
+  test("workspace-error") {
+    for {
+      _ <- server.initialize(
+        s"""/metals.json
+           |{
+           |  "a": {}
+           |}
+           |/a/src/main/scala/a/Main.scala
+           |package a
+           |object Stuff {
+           |  val stuff = "you guessed it, stuff... but with no ending!
+           |  System.exit(0)
+           |}
+           |""".stripMargin
+      )
+      _ <- server.didOpen("a/src/main/scala/a/Main.scala")
+      result <- server
+        .startDebuggingUnresolved(
+          new DebugFullyUnresolvedParams(
+            s"file://${workspace}/a/src/main/scala/a/Main.scala",
+            "run"
+          ).toJson
+        )
+        .recover { case WorkspaceErrorsException =>
+          WorkspaceErrorsException
+        }
+    } yield assertNoDiff(
+      result.toString,
+      WorkspaceErrorsException.toString()
+    )
+  }
+
+  test("testFile") {
+    for {
+      _ <- server.initialize(
+        s"""/metals.json
+           |{
+           |  "a": {
+           |    "libraryDependencies":["org.scalatest::scalatest:3.0.5"]
+           |  }
+           |}
+           |/a/src/main/scala/a/Foo.scala
+           |package a
+           |class Foo extends org.scalatest.FunSuite {
+           |  test("foo") {}
+           |}
+           |""".stripMargin
+      )
+      _ <- server.didOpen("a/src/main/scala/a/Foo.scala")
+      debugger <- server.startDebuggingUnresolved(
+        new DebugFullyUnresolvedParams(
+          s"file://${workspace}/a/src/main/scala/a/Foo.scala",
+          "testFile"
+        ).toJson
+      )
+      _ <- debugger.initialize
+      _ <- debugger.launch
+      _ <- debugger.configurationDone
+      _ <- debugger.shutdown
+      output <- debugger.allOutput
+    } yield assert(output.contains("All tests in a.Foo passed"))
+  }
```

Comment:
So I'm curious if you have any input here, I'm having a really hard time figuring out how to get this test to pass. Locally with this same exact layout everything works as expected, but when testing here the testClasses never actually get cached, so when I look for them, it just returns an empty list and therefore no tests are found. I had the same issue with testing `run`, but was able to get around it by just including the `didOpen` on the file, which seemed to ensure all the build target classes were fetched and main classes stored. However, that's not working for tests here and I'm unsure how to do this. Any ideas?

Reply:
I think we need to have the files fully compiled, so maybe add `didSave` after did open? I think it should actually wait for compilation to finish.

Reply:
Yea even with the `didSave` it still doesn't fine them 🤔

Reply:
I started looking into how we made it work in other tests and it seems we didn't! We don't actually have any tests for testing. There is also an issue I encountered in Bloop which would make the tests endpoint timeout pretty often and is the number one flaky test currently. I think it could all be connected :/

Reply:
Hmm, bummer. I'll play around a bit more with them to see if I can figure out what's going on.

Reply:
We can skip testing it currently or await some time for the tests classes information to show up. I will try to investigate the Bloop side.